code/workspaces/infrastructure-api/src/routers/clustersRouter.js

```diff
@@ -0,0 +1,27 @@
+import { service } from 'service-chassis';
+import express from 'express';
+import { permissionTypes } from 'common';
+import clustersController from '../controllers/clustersController';
+import permissionMiddleware from '../auth/permissionMiddleware';
+import { paramProjectKeyValidator } from './commonValidators';
+
+const { projectPermissions: { PROJECT_KEY_STACKS_CREATE, PROJECT_KEY_STACKS_LIST } } = permissionTypes;
+
+const { errorWrapper } = service.middleware;
+
+const clustersRouter = express.Router();
+
+clustersRouter.post(
+  '/',
+  permissionMiddleware(PROJECT_KEY_STACKS_CREATE),
+  clustersController.clusterValidator(),
+  errorWrapper(clustersController.createCluster),
+);
+clustersRouter.get(
+  '/project/:projectKey',
+  paramProjectKeyValidator(),
+  permissionMiddleware(PROJECT_KEY_STACKS_LIST),
```

Comment:
I think the permission check should come before the validation. The `permissionMiddleware` doesn't use the validated data for its checks so the validator doesn't need to have been run first.

Reply:
Good point